### PR TITLE
Mock service leveraging kuksa client lib

### PR DESCRIPTION
**Pull request by lukasmittag**
This PR leverages kuksa-client lib instead of pure grpc and protobuf files. It fixes too the EventTrigger on values.

### CODE REVIEW

**Comment by SebastianSchildt** [comment]
@doosuu  can you check whether this works/seems reasonable

File: mock_service/README.md [hunk 1]
```diff
@@ -68,3 +68,7 @@ API documentation is generated from Python docs and embedded into markdown files
 ```bash
 ./update-api-docs.sh
 ```
+
+# What's not supported?
+Any form of array support e.g no VSS datapoint that has DataType.*ARRAY. 
+If you set something not through mock it won't get pulled to the mock service. Instead use another behavior that listens to ACTUATOR_TARGET/VALUE and set it like this.
```
Comment (doosuu) [inline mock_service/README.md:74]:
```suggestion
If values in Kuksa Databroker are modified outside of behaviors, the changes won't be picked up by the mock service. Instead try to model another behavior that listens to ACTUATOR_TARGET/VALUE and set it like this.
```

File: mock_service/lib/action.py [hunk 2]
```diff
@@ -124,5 +126,5 @@ def execute(
             self._resolved_value = self._target_value_resolver(
                 action_context, self._value
             )
-
-        action_context.datapoint.set_value(self._resolved_value)
+        if self._resolved_value is not None:
+            action_context.datapoint.set_value(self._resolved_value)
```
Comment (doosuu) [inline mock_service/lib/action.py:130]: Missing empty line. Did you run pre-commit?

File: mock_service/lib/behavior.py [hunk 3]
```diff
@@ -21,6 +21,17 @@
 from lib.types import Event, ExecutionContext
 
 log = logging.getLogger("behavior")
+log.setLevel(logging.DEBUG)
+
+# Create a file handler and set the log file name
+file_handler = logging.FileHandler("mock_service.log")
```
Comment (doosuu) [inline mock_service/lib/behavior.py:27]: Do we really want to set this only for behavior.py?

File: mock_service/lib/behavior.py [hunk 4]
```diff
@@ -77,17 +88,18 @@ def execute(self, delta_time: float, animators):
...
-                ):
-                    log.info(f"Running behavior for {path}")
-                    behavior.execute(
-                        ActionContext(
-                            trigger_result, execution_context, matched_datapoint
-                        ),
-                        animators,
-                    )
-                    break
+                if behavior.is_condition_fulfilled(
```
Comment (doosuu) [inline mock_service/lib/behavior.py:93]: Is there any specific reason (i.e. a bug) that lead to the swapping of the condition check and trigger activation check?

File: mock_service/lib/datapoint.py [hunk 5]
```diff
@@ -33,11 +33,12 @@ def __init__(
         self.is_mocked = is_mocked
         self.value_listener = value_listener
 
+    # Do we need this -> never used as of now
```
Comment (doosuu) [inline mock_service/lib/datapoint.py:36]: This check **was** intended to be used whenever you find an AnimationAction on a discrete value type and throw an error. Probably this slipped through when I released the initial version.

File: mock_service/lib/dsl.py [hunk 6]
```diff
@@ -12,12 +12,15 @@
...
+from typing import Any, Callable, Dict, List, Optional
 
 from lib.action import Action, ActionContext, AnimationAction, SetAction
 from lib.animator import RepeatMode
 from lib.behavior import Behavior, ExecutionContext
-from lib.trigger import EventTriggerResult, Trigger
+from lib.trigger import EventTriggerResult, Trigger, EventType, EventTrigger
+
+# Set the log level to suppress log messages because we call connect/disconnect of client quite often
+logging.getLogger("kuksa_client").setLevel(logging.WARNING)
```
Comment (doosuu) [inline mock_service/lib/dsl.py:23]: A modification of a completely unrelated logger should definitely not be located in the public DSL module. Move it to mockservice.py?

File: mock_service/lib/dsl.py [hunk 7]
```diff
@@ -72,9 +75,12 @@ def get_datapoint_value(context: ExecutionContext, path: str, default: Any = 0)
     """
     if path not in _mocked_datapoints:
         _required_datapoint_paths.append(path)
-
-    if path in context.datapoints_cache:
-        return context.datapoints_cache[path].value
+    context.client.connect()
```
Comment (doosuu) [inline mock_service/lib/dsl.py:78]: Not sure if connecting and disconnecting is better than caching the values. Was this change driven by a bug?

File: mock_service/lib/dsl.py [hunk 8]
```diff
@@ -98,9 +104,14 @@ def __resolve_value(action_context: ActionContext, value: Any) -> Any:
 
     if isinstance(value, str) and value.startswith("$"):
         if value == "$self":
-            return get_datapoint_value(
-                action_context.execution_context, action_context.datapoint.path
-            )
+            action_context.execution_context.client.connect()
```
Comment (doosuu) [inline mock_service/lib/dsl.py:107]: What's the reason for this change? Calling get_datapoint_value would save duplicate code.

File: mock_service/lib/dsl.py [hunk 9]
```diff
@@ -109,8 +120,13 @@ def __resolve_value(action_context: ActionContext, value: Any) -> Any:
                     f"Unsupported literal: {value!r} in non event-triggered behavior!"
                 )
         elif value.startswith("$"):
-            return get_datapoint_value(action_context.execution_context, value[1:])
-
+            action_context.execution_context.client.connect()
```
Comment (doosuu) [inline mock_service/lib/dsl.py:123]: What's the reason for this change? Calling get_datapoint_value would save duplicate code.

File: mock_service/lib/dsl.py [hunk 10]
```diff
@@ -143,3 +159,18 @@ def create_animation_action(
         AnimationAction: The created AnimationAction.
     """
     return AnimationAction(duration, repeat_mode, values, __resolve_value)
+
+def create_EventTrigger(type: EventType, path: Optional[str] = None) -> EventTrigger:
```
Comment (doosuu) [inline mock_service/lib/dsl.py:163]: Does not follow established naming conventions: should be called `create_event_trigger`

File: mock_service/lib/dsl.py [hunk 11]
```diff
@@ -143,3 +159,18 @@ def create_animation_action(
...
+    """Create an EventTrigger for own VSS path or different one. It handles that events for the new VSS paths
```
Comment (doosuu) [inline mock_service/lib/dsl.py:164]:
```suggestion
    """Create an EventTrigger for the mocked datapoint in context of this call OR the explicitly passed one. 
```

File: mock_service/lib/dsl.py [hunk 12]
```diff
@@ -143,3 +159,18 @@ def create_animation_action(
...
+    are handled too and the EventTrigger works fine.
```
Comment (doosuu) [inline mock_service/lib/dsl.py:165]:
```suggestion
```

File: mock_service/lib/dsl.py [hunk 13]
```diff
@@ -143,3 +159,18 @@ def create_animation_action(
...
+
+    Args:
+        type (EvenType): The kind of event the EventTrigger shall be.
```
Comment (doosuu) [inline mock_service/lib/dsl.py:168]:
```suggestion
        type (EventType): The type of event which will activate the trigger. 
```

File: mock_service/lib/dsl.py [hunk 14]
```diff
@@ -143,3 +159,18 @@ def create_animation_action(
...
+        path (Optional[str]): Default None which represents same VSS path as the datapoint or new VSS path
```
Comment (doosuu) [inline mock_service/lib/dsl.py:169]:
```suggestion
        path (Optional[str]): The data point which shall raise the event. If not set defaults to the mocked data point in context of the call. 
```

File: mock_service/lib/types.py [hunk 15]
```diff
@@ -28,6 +32,6 @@ class ExecutionContext(NamedTuple):
     """Context in which behaviors are executed"""
 
     calling_signal_path: str
-    pending_event_list: List[Event]
-    datapoints_cache: Dict[str, MockedDataPoint]
+    pending_event_list: List[Event] 
     delta_time: float
+    client: VSSClient = VSSClient(vdb_split[0], vdb_split[1])
```
Comment (doosuu) [inline mock_service/lib/types.py:37]: Should be passed as dependency injection.

File: mock_service/mock.py
Comment (doosuu) [inline mock_service/mock.py:1]: Why were most of the wiper mocks removed from the file?

**Review by doosuu — APPROVED** [review]

**Review by doosuu — CHANGES_REQUESTED** [review]
Please rework findings.

File: mock_service/lib/action.py (hunk 2 above)
Comment (lukasmittag) [reply mock_service/lib/action.py:130]: Ah n, did not run pre-commit :)

File: mock_service/lib/behavior.py (hunk 3 above)
Comment (lukasmittag) [reply mock_service/lib/behavior.py:27]: seemed to me to have the best output but yeah maybe could add it to the other ones too.

File: mock_service/mock.py
Comment (lukasmittag) [reply mock_service/mock.py:1]: No particular reason will change back

File: mock_service/lib/dsl.py (hunk 7 above)
Comment (lukasmittag) [reply mock_service/lib/dsl.py:78]: yeah not sure too about this, but seemed like a better option at the point I made the change. For me it would be now okay to revert to caching if you think it is faster/better.

File: mock_service/lib/dsl.py (hunk 8 above)
Comment (lukasmittag) [reply mock_service/lib/dsl.py:107]: No, will change

File: mock_service/lib/behavior.py (hunk 3 above)
Comment (lukasmittag) [reply mock_service/lib/behavior.py:27]: I think for the other ones logging in log file is not necessary. Did this for seeing logs in the GUI. I remember now.

File: mock_service/lib/types.py (hunk 15 above)
Comment (lukasmittag) [reply mock_service/lib/types.py:37]:
```
class ExecutionContext(NamedTuple):
    """Context in which behaviors are executed"""

    def __init__(self, vss_client:VSSClient = VSSClient(vdb_split[0], vdb_split[1])):
        self.vss_client = vss_client
        self.calling_signal_path: str
        self.pending_event_list: List[Event] 
        self.delta_time: float
       ```
so something like this?

**Comment by lukasmittag** [comment]
CI failing due to reformatting of docs with pre-commit hooks.

File: mock_service/lib/behavior.py (hunk 3 above)
Comment (doosuu) [reply mock_service/lib/behavior.py:27]: Long term I think it might be confusing that just `behavior.py` is writing to `mock_service.log`. Maybe you want to create this file in a central location and only let behavior.py write to it with default `INFO` level?

If it serves a purpose, it seems OK for now.

File: mock_service/lib/behavior.py (hunk 4 above)
Comment (doosuu) [reply mock_service/lib/behavior.py:93]: You are right, all conditions which listen to the same event in the same iteration should get triggered. We should raise a bug issue for this. For now I understand why this change is necessary.

File: mock_service/lib/dsl.py (hunk 7 above)
Comment (doosuu) [reply mock_service/lib/dsl.py:77]: Without numbers its hard to say which one is better. You are trading runtime performance vs memory usage. Biggest downside I see is the number of log messages you are generating by connecting/disconnecting. Can the client somehow stay connected?

File: mock_service/lib/types.py (hunk 15 above)
Comment (doosuu) [reply mock_service/lib/types.py:37]: I would even remove the default initializer completely that way you maybe can share the same VSS client and get away without (re-)connecting? Since ExecutionContext is already a NamedTuple, you do not need the extra constructor.

**Review by doosuu — COMMENTED** [review]
Please have a look at my latest comments

File: mock_service/lib/dsl.py (hunk 7 above)
Comment (lukasmittag) [reply mock_service/lib/dsl.py:77]: Yes we can give over the client from baseservice then we have just on client that is connected at the beginning and then serves for all calls made to databroker. Will change it to this

File: mock_service/lib/behavior.py (hunk 4 above)
Comment (lukasmittag) [reply mock_service/lib/behavior.py:87]: #91

**Comment by doosuu** [comment]
LGTM

**Review by SebastianSchildt — APPROVED** [review]
I like
image
